tests/unit/s2n_ktls_test_utils_test.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "testlib/s2n_ktls_test_utils.h"
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "utils/s2n_random.h"
+
+#define S2N_TEST_RECORD_TYPE 43
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* test data */
+    uint8_t test_data[S2N_TLS_MAXIMUM_FRAGMENT_LENGTH] = { 0 };
+    struct s2n_blob test_data_blob = { 0 };
+    EXPECT_SUCCESS(s2n_blob_init(&test_data_blob, test_data, sizeof(test_data)));
+    EXPECT_OK(s2n_get_public_random_data(&test_data_blob));
+
+    /* Test the mock IO stuffer implementation can send/recv records */
+    {
+        /* Test sending send/recv 0 bytes */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_ktls_io_pair io_pair = { 0 },
+                    s2n_ktls_io_pair_free);
+            EXPECT_OK(s2n_test_init_ktls_stuffer_io(server, client, &io_pair));
+
+            size_t send_zero = 0;
+            /* Init send msghdr */
+            struct iovec send_msg_iov = { .iov_base = test_data, .iov_len = send_zero };
+            struct msghdr send_msg = { .msg_iov = &send_msg_iov, .msg_iovlen = 1 };
+            /* sendmsg */
+            ssize_t bytes_written = s2n_test_ktls_sendmsg_stuffer_io(server, &send_msg, S2N_TEST_RECORD_TYPE);
+            EXPECT_EQUAL(bytes_written, send_zero);
+
+            EXPECT_EQUAL(io_pair.client_in.send_recv_msg_invoked_count, 1);
+            EXPECT_EQUAL(io_pair.server_in.send_recv_msg_invoked_count, 0);
+        };
+
+        /* Test send/recv non-zero values. Zero is a special case and tested above */
```

Comment:
Does this cover all the cases you've considered in the implementation? What about reading multiple records, or blocking on a read?

Your recv mock is very complex, and I'm not convinced that it's fully tested here.

Reply:
When we originally spoke we decided to keep tests for this to a minimal. As this impl is complex enough I am going to move some of the upcoming tests to this PR

Reply:
The tests additional tests (recv + send) are adding alot of new code to a PR that is mostly reviewed. I am inclined to add the tests in a followup PR for this reason. Since this code is only used for testing anyway this should be ok.